### PR TITLE
UCP/RNDV: pack rkey together with memory domain id

### PR DESCRIPTION
**Pull request by hoopoepg**
- added info about memory domain & rkey size
- update fallback from HW tag offloading to SW rndv
- updated is_get_op_possible calls to check ability
  to rndv/get more accurate

NOTE: this PR will conflict with https://github.com/openucx/ucx/pull/2008, let's merge https://github.com/openucx/ucx/pull/2008 first

### CODE REVIEW

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5261/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5269/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5277/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5283/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by hoopoepg** [comment]
@MattBBaker could you look what is wrong?

File: src/ucp/tag/rndv.c [hunk 1]
```diff
@@ -44,26 +69,86 @@ static void ucp_rndv_rma_request_send_buffer_dereg(ucp_request_t *sreq)
...
+    int i;
+    ucp_rndv_rkey_data_t *rkey;
+
+    ucs_assert_always(cnt < UCP_MAX_RNDV_LANES);
+    ucs_assert(req->send.rndv_get.rkey != NULL);
+
+    buf++; /* first byte is counter of packet keys */
+    req->send.rndv_get.num_lanes = 0;
+
+    for (i = 0; i < cnt; i++) {
```
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: datatype mismatch between `i` and `cnt`, it's probably reason of ORNL failure. also minor - it's better to reorder variables on stack from big to small.

File: src/ucp/tag/rndv.c [hunk 2]
```diff
@@ -348,19 +433,31 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_proto_progress_rndv_get_zcopy, (self),
         return UCS_INPROGRESS;
     }
 
-    rsc_index = ucp_ep_get_rsc_index(rndv_req->send.ep, rndv_req->send.lane);
+    lane_idx = rndv_req->send.rndv_get.lane_idx;
+    lane = ucp_tag_rndv_get_lane_info(rndv_req, lane_idx)->lane;
+    rkey = ucp_tag_rndv_rkey(rndv_req, lane_idx);
+
+    rndv_req->send.rndv_get.lane_idx = (lane_idx + 1) %
```
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: maybe to separate in a special function? we may want to have more complex distribution in future.

File: src/ucp/tag/rndv.c (hunk 1 above)
Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: > datatype mismatch between i and cnt, it's probably reason of ORNL failure
nop, there is another issue, fix is in progress
> also minor - it's better to reorder variables on stack from big to small
really strange rule :)
actually most of the variables will not be located at stack (have you seen debug message about variable value is optimized?), or will be located in order which is most performance gain

File: src/ucp/tag/rndv.c [hunk 3]
```diff
@@ -385,9 +482,7 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_proto_progress_rndv_get_zcopy, (self),
     state = rndv_req->send.state.dt;
     ucp_dt_iov_copy_uct(iov, &iovcnt, max_iovcnt, &state, rndv_req->send.buffer,
                         ucp_dt_make_contig(1), length);
-
-    lane = rndv_req->send.lane;
-    rkey = ucp_tag_rndv_rkey(rndv_req, 0);
+    iov[0].memh = ucp_tag_rndv_get_lane_info(rndv_req, lane_idx)->memh;
```
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: How about to encapsulate `memh` initialization into `ucp_dt_iov_copy_uct` ?

File: src/ucp/tag/rndv.c [hunk 4]
```diff
@@ -886,7 +979,7 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_rtr_handler,
 
         UCS_PROFILE_CALL(uct_rkey_unpack, rndv_rtr_hdr + 1, &rkey_bundle);
 
-        if (!(ucp_tag_rndv_is_get_put_op_possible(ep, ucp_ep_get_rndv_get_lane(ep, 0),
+        if (!(ucp_tag_rndv_is_put_op_possible(ep, ucp_ep_get_rndv_get_lane(ep, 0),
                                                   rkey_bundle.rkey))) {
```
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: alignment

Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: will fix

File: src/ucp/tag/rndv.c (hunk 3 above)
Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: ```ucp_dt_iov_copy_uct``` is too general and not designed to take separate memh. original code (before ```ucp_dt_iov_copy_uct``` refactoring) fill single IOV entry and was easy to read, for now we have to post-process ```ucp_dt_iov_copy_uct``` call

File: src/ucp/tag/rndv.c (hunk 2 above)
Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: let's separate it after this complex distribution is defined :)

File: src/ucp/tag/rndv.c (hunk 1 above)
Comment (evgeny-leksikov) [reply src/ucp/tag/rndv.c]: `int i;`
...
`uint8_t  cnt = *buf;`

> really strange rule

ok, it's probably out-of-date rule and my personal preference :)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5290/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
one more note: there is no split of original message by lanes in this PR, it will be added in next PR

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5295/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5299/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5306/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5308/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5316/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@brminich could you look at tag-offloading part?

File: src/ucp/tag/offload.c [hunk 5]
```diff
@@ -175,9 +193,12 @@ void ucp_tag_offload_rndv_cb(uct_tag_context_t *self, uct_tag_t stag,
     rkey_size = (sw_hdr->flags & UCP_RNDV_RTS_FLAG_PACKED_RKEY) ?
                 ucp_tag_offload_rkey_size(ctx) : 0;
 
-    rts = alloca(sizeof(*rts) + rkey_size);
+    pkey_size = (sw_hdr->flags & UCP_RNDV_RTS_FLAG_PACKED_RKEY) ?
```
Comment (brminich) [inline src/ucp/tag/offload.c]: seems like better have one common `if` for both `rkey_size` and `pkey_size`

File: src/ucp/tag/offload.c [hunk 6]
```diff
@@ -201,15 +222,15 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_tag_offload_unexp_rndv,
     if (remote_addr) {
         /* Unexpected tag offload RNDV */
         ucp_tag_offload_fill_rts(rts, rndv_hdr, stag, remote_addr, length, 0);
-        len = ucp_tag_offload_copy_rkey(rts, (void*)rkey_buf, rkey_size) +
+        len = ucp_tag_offload_copy_rkey(worker->context, rts, (void*)rkey_buf, rkey_size) +
```
Comment (brminich) [inline src/ucp/tag/offload.c]: it is incorrect, since rts allocated with `sizeof(*rts) + rkey_size` length and you need more memory in `ucp_tag_offload_copy_rkey`

File: src/ucp/tag/rndv.c [hunk 7]
```diff
@@ -29,6 +29,88 @@ static int ucp_tag_rndv_is_get_put_op_possible(ucp_ep_h ep, ucp_lane_index_t lan
...
+    ucp_tag_rndv_resolve_lanes(req);
+
+    if (req->send.rndv_get.num_lanes == 0) {
+        return 0;
+    }
+
+    for (i = 0; i < req->send.rndv_get.num_lanes; i++) {
+        lane = ucp_tag_rndv_get_lane_info(req, i)->lane;
+        md_flags = ucp_ep_md_attr(ep, lane)->cap.flags;
+        if ((md_flags & UCT_MD_FLAG_NEED_RKEY) && !ucp_tag_rndv_is_rkey_valid(req, i)) {
```
Comment (brminich) [inline src/ucp/tag/rndv.c]: you already checked that in `ucp_tag_rndv_resolve_lanes`. Seems no need on this check

File: src/ucp/tag/rndv.c [hunk 8]
```diff
@@ -44,26 +126,77 @@ static void ucp_rndv_rma_request_send_buffer_dereg(ucp_request_t *sreq)
 
 static size_t ucp_tag_rndv_pack_send_rkey(ucp_request_t *sreq, void *rkey_buf, uint16_t *flags)
 {
-    ucp_ep_h ep = sreq->send.ep;
-    ucp_lane_index_t lane = ucp_ep_get_rndv_get_lane(ep, 0);
+    ucp_ep_h ep           = sreq->send.ep;
```
Comment (brminich) [inline src/ucp/tag/rndv.c]: minor: no need in so big indentation

File: src/ucp/tag/rndv.c [hunk 9]
```diff
@@ -44,26 +126,77 @@ static void ucp_rndv_rma_request_send_buffer_dereg(ucp_request_t *sreq)
...
 
     ucs_assert(UCP_DT_IS_CONTIG(sreq->send.datatype));
 
-    /* Check if the sender needs to register the send buffer -
-     * is its datatype contiguous and does the receive side need it */
-    if (ucp_ep_rndv_md_flags(ep, 0) & UCT_MD_FLAG_NEED_RKEY) {
-        status = ucp_request_rndv_buffer_reg(sreq);
-        ucs_assert_always(status == UCS_OK);
+    *cnt   = 0;
+    packet = 1; /* packet at least counter */
```
Comment (brminich) [inline src/ucp/tag/rndv.c]: do you mean `packet_len` or `packed`? Just `packet` looks misleading

File: src/ucp/tag/rndv.c [hunk 10]
```diff
@@ -29,6 +29,88 @@ static int ucp_tag_rndv_is_get_put_op_possible(ucp_ep_h ep, ucp_lane_index_t lan
...
+        if (info->lane != UCP_NULL_LANE) {
+            i++;
+        } else {
+            /* could not find pair lane - release rkey */
+            if (ucp_tag_rndv_is_rkey_valid(req, i)) {
+                uct_rkey_release(ucp_tag_rndv_rkey_bundle(req, i));
+            }
+            /* shift rest of rkeys to current location */
+            if (i + 1 < req->send.rndv_get.num_lanes) {
+                memmove(info, ucp_tag_rndv_get_lane_info(req, i + 1),
```
Comment (brminich) [inline src/ucp/tag/rndv.c]: maybe it's better to have a map of matched keys then memmove?

File: src/ucp/tag/rndv.c [hunk 11]
```diff
@@ -457,15 +598,29 @@ static void ucp_rndv_handle_recv_contig(ucp_request_t *rndv_req, ucp_request_t *
...
+        if (!ucp_ep_is_stub(rndv_req->send.ep)) {
+            /* for regular, not stub, ep's we may resolve lanes here and
+             * set wakeup lane from resolved lanes */
+            ucp_tag_rndv_resolve_lanes(rndv_req);
+            rndv_req->send.lane = ucp_tag_rndv_get_lane_info(rndv_req, 0)->lane;
+        } else {
+            rndv_req->send.lane = UCP_NULL_LANE;
+        }
+
+        if (rndv_req->send.lane == UCP_NULL_LANE) {
```
Comment (brminich) [inline src/ucp/tag/rndv.c]: this is not needed now, right? Because `ucp_tag_rndv_resolve_lanes` should find the proper lane in either way (offload or not). BTW, if you remove this condition branch, make sure you don't assign `rndv_req->send.lane` to `UCP_NULL_LANE` in case of stub ep, otherwise you will get segfault while adding request to the pending queue on UCT ep.

File: src/ucp/tag/rndv.c (hunk 10 above)
Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: bit optimized: now it copy last element to current position

File: src/ucp/tag/rndv.c (hunk 8 above)
Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: fixed

File: src/ucp/tag/rndv.c (hunk 9 above)
Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: renamed

File: src/ucp/tag/rndv.c (hunk 11 above)
Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: incorrect: in case if no one matched line is found - then pre-defined tag or rndv lane is used to ensure that progress works properly (to call progress callback). in this case rndv will fallback to AM mode

File: src/ucp/tag/rndv.c (hunk 7 above)
Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: resolve_lanes just looks for corresponding lane, it doesn't check for flags & rkey valid value. in case if something wrong (for example tag offloading failed to provide rkey in future) - rndv will safely fallback to AM mode

Comment (brminich) [reply src/ucp/tag/rndv.c]: can't we do that in `ucp_tag_rndv_resolve_lanes`? Otherwise why would we need to add lanes which can't be used for rndv?

File: src/ucp/tag/rndv.c (hunk 11 above)
Comment (brminich) [reply src/ucp/tag/rndv.c]: in this case it is worth to use am lane regardless of `UCP_RNDV_RTS_FLAG_OFFLOAD` presence

File: src/ucp/tag/rndv.c (hunk 7 above)
Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: ok, fixed

File: src/ucp/tag/rndv.c (hunk 11 above)
Comment (brminich) [reply src/ucp/tag/rndv.c]: why? In this case we have just one lane capable of AM bcopy. When stub ep is switched to normal one, all pending operations are moved to that ep. You may check `ucp_ep_create_stub`

**Comment by hoopoepg** [comment]
bot:bgate:retest

**Comment by hoopoepg** [comment]
bot:ornl:retest

File: src/ucp/tag/rndv.c (hunk 11 above)
Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: yes, but stub EP is just one of the corner case, current implementation is the simplest and covers all possible cases (stub, no lanes found, tag offload falled back to rndv-get, etc)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5330/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5331/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5336/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5341/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
will use ucp_rkey instead
